CAMEL-13082: Filter results already seen by the Olingo4 consumer

phantomjinx opened the pull request:
* Olingo4Configuration
 * Adds filterAlreadySeen property
 * Added here rather than to the endpoint as the latter would require
   ultimately, in order to satisfy various checks, changes to the
   endpoint read() method in the api's Olingo4AppImpl, which is
   unnecessary since this is used by the consumer.

* Olingo4Endpoint
 * Adds filterAlreadySeen property names to the property names collection
   so that parseQueryParams() does not remove it.
 * Adds the filterAlreadySeen property during interceptProperties() in a
   just-in-time manner (avoiding possibility of it causing problems with
   endpoint and consumer initialisation).

* Olingo4Consumer
 * Adds a resultIndex for indexing hashcodes of existing results
 * On receiving a response, if the resultIndex has been initialised then
   the response results are checked by getting their hashcodes. If they're
   in the resultIndex then they are discarded.
 * Prior to any polling interceptProperties is called and if the properties
   contain the filterAlreadySeen property then the resultIndex is initialised.
 * Once the results have been placed in the exchange, interceptResult() is
   called and some housekeeping is done to index all the results in the
   resultIndex ready for checking on the next polling.

dmvolod commented:
/lgtm, but would be nice to have a test for this feature as we are using OData 4.0 based cloud service for tests.

phantomjinx commented:
> /lgtm, but would be nice to have a test for this feature as we are using OData 4.0 based cloud service for tests.

Have ...
- updated olingo2 component as well as olingo4
- done tests for both components

Will update this PR imminently.

dmvolod commented:
@phantomjinx did you run full tests on olingo2 and olingo4 components? I run them on PR and fails with errors locally.

phantomjinx commented:
@dmvolod 
I did but did not fetch first. Something has changed hence the failing tests. Anyway, I have rebased, fixed the tests and updated the PR. Hopefully, you shouldn't see any more test failures. If there are then please let me know.
